### PR TITLE
Improves lazy/eager, bundle engine & roadmap documentation

### LINKED ISSUE #56: We need clarify roadmap & lazy/eager engines
1) Clarify that engines should always be built together with their host app and other engines, regardless of whether they are lazy or eager. This is required for hosts to understand and properly bundle engine dependencies and routes.

2) Update the "Engines Roadmap" section with  -  https://discuss.emberjs.com/t/engines-1-0-roadmap/14914

### PR DESCRIPTION
**Pull request by villander**
it's fixed #56 

cc: @dgeb @MelSumner

### CODE REVIEW

**Review by MelSumner — APPROVED** [review]
Thank you for doing this!!

**Comment by dgeb** [comment]
@villander thank you for putting this together!

I've begun reviewing. Please do not merge this yet.

File: markdown/guide/addons.md [hunk 1]
```diff
@@ -23,6 +23,16 @@ ember-engines will automatically try to deduplicate addons used by your host app
...
+Isolation and liability is generally hard to solve, all possible cases have down sides here:
+
+ 1) Include shared deps into the host app - Downside: Increases initial bundle size.
+
+ 2) Include shared deps into each engine bundle - Downside: Duplicated stuff in multiple engine bundles.
+
+ 3) Somehow split out shared deps into dedicated bundles `(e.g. /engine-assets/shared-by-a-and-b.js)` - Downside: You might end up with 20 bundles `(e.g. shared-by-a-and-b, shared-by-b-and-c-, shared-by-a-b-and-c, ...)`, and it is probably pretty complicated from a built step perspective
+
+The **case 1** is currently the best approach
+
```
Comment (dgeb) [inline markdown/guide/addons.md:29]: IMO, the above section doesn't really work following the prior two paragraphs, which explain what ember-engines actually does. It would make more sense as a lead-in to the problem that starts the section.

Also, saying "The **case 1** is currently the best approach" isn't particularly useful. This could be dropped in favor of describing the approach that ember-engines takes and why.

File: markdown/guide/addons.md [hunk 2]
```diff
@@ -23,6 +23,16 @@ ember-engines will automatically try to deduplicate addons used by your host app
 
 If addons are not included in the host app, they will be included in the engine's vendor bundle. 
 
+Isolation and liability is generally hard to solve, all possible cases have down sides here:
```
Comment (dgeb) [inline markdown/guide/addons.md]: "and liability"?

File: markdown/guide/creating-an-engine.md [hunk 3]
```diff
@@ -9,9 +9,10 @@ As discussed previously, Engines are distributed as Addons and In-Repo-Addons.
 To create an engine within an existing application's project, run the
 `in-repo-engine` generator:
 
-```
+```bash
 ember g in-repo-engine <engine-name>
 ```
+> Note: In-repo addons currently are created in the `/lib` directory (e.g. `/lib/my-addon`). Unit tests and integration tests are currently co-mingled with tests for the main application. It's recommended that in-repo engines provide better test separation than is provided for regular addons, and perhaps the whole in-repo addon directory structure should be re-examined at the same time in-repo engines are introduced.
```
Comment (dgeb) [inline markdown/guide/creating-an-engine.md]: > It's recommended that in-repo engines provide better test separation than is provided for regular addons, and perhaps the whole in-repo addon directory structure should be re-examined at the same time in-repo engines are introduced.

I don't understand why this sentence should be in a Quickstart guide to creating an engine.

File: markdown/guide/creating-an-engine.md [hunk 4]
```diff
@@ -23,6 +24,25 @@ This has added a directory named `lib` and an app addon directory structure name
 
 Using the blueprint in-repo-engine, ember-cli has added all the appropriate files to create a new app structure.
 
+If you want to do is to depend on the `common` in-repo addon, so we can re-use the common elements within the engine. Let’s take a look at `lib/super-blog/package.json`:
```
Comment (dgeb) [inline markdown/guide/creating-an-engine.md]:
```suggestion
If you want to share elements between an in-repo engine and application, you could do so with an in-repo addon. For example, the following in-repo engine's package (`lib/super-blog/package.json`) references the `common` in-repo addon:
```

File: markdown/guide/lazy-loading.md [hunk 5]
```diff
@@ -57,24 +57,161 @@ Note that route serializers are unique to Engines and won't work in normal appli
...
+`import`s will also be in the `lazy-engine` module path.
+
+ #### `/app`
+
+ Assets in this folder don't make sense and will be ignored as they break the
+isolation guarantees of engines.
+
+ #### `/addon`
+
+ JavaScript assets in this folder will be processed as per normal addon behavior
```
Comment (dgeb) [inline markdown/guide/lazy-loading.md]:
```suggestion
JavaScript assets in this folder will be processed as per normal addon behavior
```

File: markdown/guide/what-are-engines.md [hunk 6]
```diff
@@ -1,27 +1,43 @@
 ## What Are Engines?
```
Comment (dgeb) [inline markdown/guide/what-are-engines.md]: This is a massive rewrite of this section that would be better presented in a separate PR, or multiple PRs. The messaging is very important.

File: markdown/guide/mounting-engines.md [hunk 7]
```diff
@@ -23,6 +23,18 @@ Anyway, let's add something super simple to our template:
...
+
+Engines can exist in several phases:
+
+ **Booted** - an engine that's been installed in a parent application will have its dependencies loaded and its (non-instance) initializers invoked when the parent application boots.
+
+ **Mounted** - Routable and route-less engines have slightly different concepts of "mounting". A routable engine is considered mounted when it has been included by a router at one or more mount-points. A route-less engine is considered mounted as soon as a route's `mount` call resolves.
+
+ **Instantiated** - When an engine is instantiated, an `EngineInstance` is created and an engine's instance initializers are invoked. A routable engine is instantiated when a route is visited at or beyond its mount-point. A route-less engine is instantiated as soon as it is mounted.
+
+Special `before` and `after` hooks could be added to application instance initializers that allow them to be ordered relative to engine instance initializers.
```
Comment (dgeb) [inline markdown/guide/mounting-engines.md]: Perhaps link to the ember guides for instance initializers?

File: markdown/guide/lazy-loading.md [hunk 8]
```diff
@@ -57,24 +57,161 @@ Note that route serializers are unique to Engines and won't work in normal appli
...
+file. This location should be considered `undefined` behavior and should not be
+relied upon as it may change in the future.
+
+ Its module name inside of the host application will be `lazy-engine/routes`. Any
+`import`s will also be in the `lazy-engine` module path.
+
+ #### `/app`
+
+ Assets in this folder don't make sense and will be ignored as they break the
+isolation guarantees of engines.
```
Comment (dgeb) [inline markdown/guide/lazy-loading.md:112]: I'm not positive that app assets are actually ignored. I agree it's a bad idea to use them.

File: markdown/guide/lazy-loading.md [hunk 9]
```diff
@@ -57,24 +57,161 @@ Note that route serializers are unique to Engines and won't work in normal appli
...
+The loading and error substates work in a similar fashion to [substates in a normal Ember app](https://guides.emberjs.com/release/routing/loading-and-error-substates/). The only difference is that lazily loaded Engines will enter a loading state while the assets for the Engine are loaded and can enter an error state when an asset fails to load.
 
-### Build Behavior
+### Built Engine Output
+
+### Eager Engines
+
+Eager engines are built approximately the same as existing addons. Differences
+are limited to consolidating the namespace of `app` code inside of an engine.
+Into the engine's namespace instead of the host application.
```
Comment (dgeb) [inline markdown/guide/lazy-loading.md]: sentence fragment

**Review by dgeb — CHANGES_REQUESTED** [review]
Thanks for putting this together. It's quite a bit to digest as a single PR. I've requested that you break at least one part into a separate PR, although it may make sense to break off any other areas that seem controversial / questionable as well. Either way, it's going to take several phases of review.

File: markdown/guide/addons.md (hunk 2 above)
Comment (villander) [reply markdown/guide/addons.md]: yes, ember engines helps in separating concepts and responsibilities right?

I put liability as synonym of responsibility

File: markdown/guide/creating-an-engine.md (hunk 3 above)
Comment (villander) [reply markdown/guide/creating-an-engine.md]: is not it important to talk about how an in-repo addon will work? We can remove it if it is undercut.

File: markdown/guide/addons.md (hunk 1 above)
Comment (villander) [reply markdown/guide/addons.md:29]: > IMO, a seção acima não funciona realmente seguindo os dois parágrafos anteriores, que explicam o que os motores de ember realmente fazem. Isso faria mais sentido como uma introdução ao problema que inicia a seção.

I'll improve it

sorry for scenario `Imagine the css or components that are only used by engine-a, engine-b and engine-c` the **CASE 2** is the best approach today. We talked about this in some issues in the ember engines repo.

Comment (villander) [reply markdown/guide/addons.md:29]: fixed @dgeb

File: markdown/guide/addons.md (hunk 2 above)
Comment (villander) [reply markdown/guide/addons.md]: fixed!

File: markdown/guide/what-are-engines.md (hunk 6 above)
Comment (villander) [reply markdown/guide/what-are-engines.md]: moved for next PR

File: markdown/guide/mounting-engines.md (hunk 7 above)
Comment (villander) [reply markdown/guide/mounting-engines.md]: moved for next PR

File: markdown/guide/lazy-loading.md (hunk 9 above)
Comment (villander) [reply markdown/guide/lazy-loading.md]: moved for next PR

File: markdown/guide/lazy-loading.md (hunk 8 above)
Comment (villander) [reply markdown/guide/lazy-loading.md:112]: @dgeb sure, `app` assets don't are ignored, but our documentation teaches to reexport components in the `app/`.

What do we do in this case?

File: markdown/guide/mounting-engines.md (hunk 7 above)
Comment (villander) [reply markdown/guide/mounting-engines.md]: fixed!

File: markdown/guide/lazy-loading.md (hunk 9 above)
Comment (villander) [reply markdown/guide/lazy-loading.md]: fixed!

File: markdown/guide/lazy-loading.md (hunk 8 above)
Comment (villander) [reply markdown/guide/lazy-loading.md:112]: fixed

File: markdown/guide/addons.md [hunk 10]
```diff
@@ -19,9 +19,13 @@ The easiest way to realise this is to define your models in a shared addon, whic
 
 ### Addon deduplication
 
+If you are using ember engines **v0.7.1 or bigger**: Use `EMBER_ENGINES_ADDON_DEDUPE` environment variable to deduplicate the nested addons of lazy engine which are also host app addons. More details at [#595](https://github.com/ember-engines/ember-engines/pull/595).
```
Comment (dgeb) [inline markdown/guide/addons.md]:
```suggestion
If you are using ember engines **v0.7.1 or higher**: Use `EMBER_ENGINES_ADDON_DEDUPE` environment variable to deduplicate the nested addons of lazy engine which are also host app addons. More details at [#595](https://github.com/ember-engines/ember-engines/pull/595).
```

File: markdown/guide/creating-an-engine.md (hunk 3 above)
Comment (dgeb) [reply markdown/guide/creating-an-engine.md]: I think this sentence is an action item for the engines team, and just adds confusion to someone starting a new engine. The sentence "Unit tests and integration tests are currently co-mingled with tests for the main application." covers most of it. You could follow this with a warning: "A separate repo should be used for your engine if you wish to avoid this co-mingling."

File: markdown/guide/addons.md [hunk 11]
```diff
@@ -45,6 +49,16 @@ module.exports = {
...
 
+Isolation and separation of responsibilities is generally hard to solve. The most recurrent scenario in a composable app with Ember Engines is: `Imagine the css or components that are only used by engine-a, engine-b and engine-c`  all possible cases have down sides here:
+
+ 1) Include shared deps into the host app - Downside: Increases initial bundle size.
+
+ 2) Include shared deps into each engine bundle - Downside: Duplicated stuff in multiple engine bundles.
+
+ 3) Somehow split out shared deps into dedicated bundles `(e.g. /engine-assets/shared-by-a-and-b.js)` - Downside: You might end up with 20 bundles `(e.g. shared-by-a-and-b, shared-by-b-and-c-, shared-by-a-b-and-c, ...)`, and it is probably pretty complicated from a built step perspective.
+
+The **case 2** is currently the best approach.
```
Comment (dgeb) [inline markdown/guide/addons.md]: Saying "best" here is a bit subjective. You could simply say "Engines currently take the second approach, but may allow more options in the future."

**Review by dgeb — CHANGES_REQUESTED** [review]
@villander just a couple more little tweaks, then this should be ready to merge imo.

File: markdown/guide/addons.md (hunk 11 above)
Comment (villander) [reply markdown/guide/addons.md]: fixed!

File: markdown/guide/creating-an-engine.md (hunk 3 above)
Comment (villander) [reply markdown/guide/creating-an-engine.md]: fixed @dgeb

**Review by dgeb — APPROVED** [review]
Thanks @villander - LGTM!
